### PR TITLE
Set unit to 1 if invalid

### LINKED ISSUE #4379: Enforce metric instrument unit requirements
We do no validation on instrument units, but should according to the [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#instrument-unit).

Similar to the metric name validation issue #4378, we could do a number of different things if a unit violates the rule. However, we may choose to do something different given that unit is optional and doesn't play as crucial a role in the identity of a metric. For example, we could decide to omit the unit if invalid and continue to export data - the same option isn't available for instrument name.

### PR DESCRIPTION
**Pull request by jack-berg**
Resolves #4379.

### CODE REVIEW

**Review by jkwatson — APPROVED** [review]

File: sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/AbstractInstrumentBuilder.java [hunk 1]
```diff
@@ -51,7 +58,32 @@ public BuilderT setDescription(String description) {
...
+  /**
+   * Determine if the instrument unit is valid. Units must be non-null, non-blank, strings of 63 or
+   * less ASCII characters.
+   */
+  // Visible for testing
+  static boolean isValidUnit(String unit) {
+    return unit != null
+        && !unit.equals("")
+        && unit.length() < 64
+        && StandardCharsets.US_ASCII.newEncoder().canEncode(unit);
```
Comment (anuraaga) [inline sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/AbstractInstrumentBuilder.java:86]: This probably needs clarification in the spec but I wouldn't allow control characters be allowed if restricting. Alternatively, given description is descriptive, I'm wondering why it's not a UTF-8 string.

File: sdk/metrics/src/test/java/io/opentelemetry/sdk/metrics/SdkMeterTest.java [hunk 2]
```diff
@@ -341,9 +346,53 @@ void testDoubleGauge() {
...
+    sdkMeter.gaugeBuilder("name").setUnit(unit).buildWithCallback(unused -> {});
+    sdkMeter.gaugeBuilder("name").setUnit(unit).ofLongs().buildWithCallback(unused -> {});
+
+    assertThat(instrumentBuilderLogs.getEvents())
+        .hasSize(12)
+        .allSatisfy(
+            log ->
+                assertThat(log.getMessage())
+                    .isEqualTo(
+                        "Instrument name unit \"日\" is invalid, using 1 instead. Instrument unit must be 63 or less ASCII characters."));
```
Comment (anuraaga) [inline sdk/metrics/src/test/java/io/opentelemetry/sdk/metrics/SdkMeterTest.java:377]: While the instrument name in this case is exacerbating the fact, the log message might be a bit hard to parse out. "Unit \"熊\" for instrument \"name\" is invalid"?

File: sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/AbstractInstrumentBuilder.java (hunk 1 above)
Comment (jack-berg) [reply sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/AbstractInstrumentBuilder.java:86]: Perhaps I misunderstand the comment. UTF-8 characters are allowed for description, but unit is more restrictive. Agree with point about the control characters. Opened this [issue](https://github.com/open-telemetry/opentelemetry-specification/issues/2496) in the spec to address.

**Comment by jack-berg** [comment]
I've gone ahead and added validation in `DefaultMeter` as well. This allows you to identify instrumentation with invalid unit without the SDK installed, and is symmetric to the [instrument name validation](https://github.com/open-telemetry/opentelemetry-java/pull/4383).

File: sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/AbstractInstrumentBuilder.java (hunk 1 above)
Comment (anuraaga) [reply sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/AbstractInstrumentBuilder.java]: Ooops, got unit and description mixed up

**Review by anuraaga — APPROVED** [review]
